### ISSUE TITLE
Swallowed exception in AndroidSystemFileProvider

### ISSUE THREAD

**Issue by SillieA23** [henrik-lindqvist/safs#2]
Exceptions from DocumentProviders are swallowed and no information is passed to the stack trace.

The specific block giving grief starts on line AndroidSystemFileProvider:486.

It would be useful to see the RuntimeException in case there is any useful information.

**Comment by henrik-lindqvist** [comment]
A DocumentProvider/ContentProvider can, and will, basically only throw FileNotFoundException, see:
https://github.com/aosp-mirror/platform_frameworks_base/blob/pie-release/core/java/com/android/internal/content/FileSystemProvider.java#L425

For whatever reason, in Android 9, it started rethrowing those as IllegalStateException instead for some failures, see:
https://github.com/aosp-mirror/platform_frameworks_base/blob/pie-release/packages/ExternalStorageProvider/src/com/android/externalstorage/ExternalStorageProvider.java#L588

The java.nio.file API don't/shouldn't throw RuntimeException/IllegalStateException on failing IO operations. Error reporting on Android is very lacking, silent failures, rethrowing nonsense exceptions without including the root cause (stacktrace), etc., as seen in above source files.

AndroidSystemFileProvider could instead throw an IOException with the "cause" but without the "file", or an FileSystemException without the "cause" but with the "file". However the "cause" will likely be nonsense anyway.

**Comment by SillieA23** [comment]
Yes you are totally right, the root exception I was having turned out to be a null pointer exception inside the method: android.content.ContentResolver.checkLeakDetectIgnoreList(java.lang.Throwable). 

The FileNotFoundException that was thrown was misleading in my case, as the file was created.

However knowing this was the issue, I could fix it via reflection.  This could be useful for others when they get an issue like this?

I understand that FileNotFoundException is the expected exception, and the aim of the library is to be easily interchangeable with java.nio.file. However, due to androids issues maybe it is a good idea to have some sort of UnhandledAndroidInternalExceptionHandler, which could be a parameter when the library is being initialised? 
The library could call this handler on unexpected errors so that the developers can see them. This doesn't stop the library being compatible with java.nio.file and also provides android developers with more information to fix these issues.

**Comment by henrik-lindqvist** [comment]
I'll create a new DocumentProviderException which extends FileSystemException with the reason set to the message of the RuntimeException. But as said, the DocumentProvider throws FileNotFoundException, which is converted to an NoSuchFileException, for a lot of other unrelated failures anyway.
